tensorflow/python/framework/tensor_util.py

```diff
@@ -44,8 +44,8 @@ def ExtractBitsFromFloat16(x):
 
 
 def SlowAppendFloat16ArrayToTensorProto(tensor_proto, proto_values):
-  tensor_proto.half_val.extend(
-      [ExtractBitsFromFloat16(x) for x in proto_values])
+  tensor_proto.half_val.extend(np.asarray(proto_values, dtype=np.float16).view(np.uint16))
```

Comment:
is there a unit test checking correctness of this code?

Reply:
(and is there any reason not to rewrite ExtractBitsFromFloat16 to use the np.asarray version?)

Reply:
Let's look at the logic of the original code.
1.tensor_proto.half_val.extend(input) 
input = [ExtractBitsFromFloat16(x) for x in proto_values]
So, input is obviously a list
2.what is the element in this list (input)
ExtractBitsFromFloat16(x)
bt the way, the proto_values is a np_array and the dimension is one , if the dimension is bigger than one , the element in proto_values is not np_number, it will also be a np_array, then this method is wrong. Remeber this, it is very important
3.what is the method ExtractBitsFromFloat16 doing?
np.asarray(x, dtype=np.float16).view(np.uint16).item()
this method do many things. At first, as we mentioned earlier, x is a np_number, so the init writer use the method np.asarray  to transfer the np_number to  np_array. Secondly, he/she uses the np_array.view(np.uint16) to do something change. At last, he/she uses the method np_array.item() to get the element in the array. But obviously, this np_array has just only one element is the x.

Now, what changes have I made?
At first. for the method tensor_proto.half_val.extend(input) 
this input can receive the input is a list or is a np_array which must be one-dimensional.
In my rewrite code, this input equals np.asarray(proto_values, dtype=np.float16).view(np.uint16)
this obviously is a np_array, as we mentioned earlier, because the proto_values is one-dimensional. so my input is also one-dimensional. So this input is lawful.
Secondly, let's look at the view method again. this method is for np_array, so it can transfer the whole np_array to what we want.

All in all, the logic of the original method is that:
1.picking each element from the one-dimensional np_array 
2.change the np_number to **np_array what we want**
3.change the np_array to **what we want** 
4.get the element from the np_array
5.combine these elements to a list
6. deal with this list

 the logic of the my method is that:
1.change the np_array to **np_array what we want**
2.change the np_array to **what we want** 
3.deal with this np_array

Using loops to process np elements is a time-consuming task. My code avoids using loops, so it's very fast.


So based on the above logic, if the original code is correct, then the code I changed is also correct.

By the way, I don't understand why you're concerned about using np. asarray

Reply:
Unit tests are there to ensure that future changes don't cause a regression.  It would take you less time to add one than to write that comment ;)

Reply:
OK, fine.
Can you give me a template for unit testing so that I can write this unit test?
I looked at other merge PR and found no unit tests that could be replicated. And I always thought that such a large project would automatically test the code submitted by individuals.

Reply:
you can put the test here:

https://github.com/tensorflow/tensorflow/blob/dd000bf0222797e71267e442e9aaa54e189e08cf/tensorflow/python/framework/tensor_util_test.py

see for example the function "testHalf".